### PR TITLE
Soroban metrics improvement

### PR DESCRIPTION
**Pull request by jayz22**
# Description

Resolves all the items in this checklist: https://github.com/stellar/stellar-core-internal/issues/236#issuecomment-1915049419

Most of them are just some naming and types change. There is one structural change:

A new `SorobanLedgerMetrics` was added in `LedgerManager` which tracks soroban resource usage per ledger across all soroban ops (invoke_host_fn, restore_footprint, extend_ttl), and publishes them to new metrics "soroban.ledger.xxx".

# Checklist
- [ ] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [ ] Rebased on top of master (no merge commits)
- [ ] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [ ] Compiles
- [ ] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: docs/metrics.md [hunk 1]
```diff
@@ -166,39 +166,49 @@ soroban.host-fn-op.read-entry                | meter     | number of entries rea
 soroban.host-fn-op.write-entry               | meter     | number of entries written
 soroban.host-fn-op.read-key-byte             | meter     | number of key bytes in read entries
 soroban.host-fn-op.write-key-byte            | meter     | number of key bytes in written entries
-soroban.host-fn-op.read-ledger-byte          | meter     | number of data + code bytes in read entries
+soroban.host-fn-op.read-ledger-byte          | meter     | number of entry (data + code) bytes in read entries
```
Comment (MonsieurNicolas) [inline docs/metrics.md:169]: wording is weird (it was confusing before too), should it just be number of bytes read?

File: docs/metrics.md [hunk 2]
```diff
@@ -166,39 +166,49 @@ soroban.host-fn-op.read-entry                | meter     | number of entries rea
...
 soroban.host-fn-op.read-data-byte            | meter     | number of data bytes in read entries
 soroban.host-fn-op.read-code-byte            | meter     | number of code bytes in read entries
-soroban.host-fn-op.write-ledger-byte         | meter     | number of data + code bytes in written entries
+soroban.host-fn-op.write-ledger-byte         | meter     | number of entry (data + code) bytes in written entries
```
Comment (MonsieurNicolas) [inline docs/metrics.md:172]: same wrt writes -- maybe just "number of bytes written for Soroban entries"

File: docs/metrics.md [hunk 3]
```diff
@@ -166,39 +166,49 @@ soroban.host-fn-op.read-entry                | meter     | number of entries rea
...
 soroban.host-fn-op.write-code-byte           | meter     | number of code bytes in written entries
 soroban.host-fn-op.emit-event                | meter     | number of events emitted
 soroban.host-fn-op.emit-event-byte           | meter     | number of event bytes emitted
 soroban.host-fn-op.cpu-insn                  | meter     | metered cpu instructions
 soroban.host-fn-op.mem-byte                  | meter     | metered memory bytes
-soroban.host-fn-op.invoke-time-nsecs         | meter     | time [nsecs] spent in `invoke_host_function`
-soroban.host-fn-op.cpu-insn-excl-vm          | meter     | metered cpu instructions excluding VM instantation
-soroban.host-fn-op.invoke-time-nsecs-excl-vm | meter     | time [nsecs] spent in `invoke_host_function` excluding VM instantation
+soroban.host-fn-op.invoke-time-nsecs         | timer     | time spent in `invoke_host_function`
+soroban.host-fn-op.cpu-insn-excl-vm          | meter     | metered cpu instructions excluding VM instantiation
```
Comment (MonsieurNicolas) [inline docs/metrics.md:180]: should specify -- is this per `invoke_host_function` op?

Comment (MonsieurNicolas) [reply docs/metrics.md:180]: as in, if a single `invoke_host_function`  op calls multiple VMs, is it recorded as one or multiple events?

Comment (MonsieurNicolas) [reply docs/metrics.md:180]: (we probably want 1 event per op)

File: docs/metrics.md [hunk 4]
```diff
@@ -166,39 +166,49 @@ soroban.host-fn-op.read-entry                | meter     | number of entries rea
...
+soroban.host-fn-op.invoke-time-fsecs-cpu-insn-ratio         | histogram | ratio between invoke_time (femto-seconds) and cpu instructions
+soroban.host-fn-op.invoke-time-fsecs-cpu-insn-ratio-excl-vm | histogram | ratio between invoke_time (femto-seconds) and cpu instructions excluding VM instantiation
 soroban.host-fn-op.max-rw-key-byte           | meter     | bytes of the largest key in entries read/written
 soroban.host-fn-op.max-rw-data-byte          | meter     | bytes of the largest data entry read/written
 soroban.host-fn-op.max-rw-code-byte          | meter     | bytes of the largest code entry read/written
 soroban.host-fn-op.max-emit-event-byte       | meter     | bytes of the largest event emitted
-soroban.host-fn-op.success                   | meter     | if `InvokeHostFunctionOp` results in a success
-soroban.host-fn-op.failure                   | meter     | if `InvokeHostFunctionOp` results in a failure
+soroban.host-fn-op.success                   | meter     | number of successful `InvokeHostFunctionOp` operations
+soroban.host-fn-op.count                     | meter     | number of total `InvokeHostFunctionOp` operations
```
Comment (MonsieurNicolas) [inline docs/metrics.md:189]: we don't need this metric (total count): count is already recorded in any other metric that tracks `InvokeHostFunctionOp` data, like `soroban.host-fn-op.invoke-time-nsecs`

File: docs/metrics.md [hunk 5]
```diff
@@ -166,39 +166,49 @@ soroban.host-fn-op.read-entry                | meter     | number of entries rea
...
-soroban.config.tx-max-read-ledger-entries        | meter     | soroban config setting `tx_max_read_ledger_entries`
-soroban.config.tx-max-read-bytes                 | meter     | soroban config setting `tx_max_read_bytes`
-soroban.config.tx-max-write-ledger-entries       | meter     | soroban config setting `tx_max_write_ledger_entries`
-soroban.config.tx-max-write-bytes                | meter     | soroban config setting `tx_max_write_bytes`
-soroban.config.bucket-list-target-size-bytes     | meter     | soroban config setting `bucket_list_target_size_bytes`
-soroban.config.tx-max-contract-events-size-bytes | meter     | soroban config setting `tx_max_contract_events_size_bytes`
-soroban.config.contract-data-key-size-bytes      | meter     | soroban config setting `contract_data_key_size_bytes`
-soroban.config.contract-data-entry-size-bytes    | meter     | soroban config setting `contract_data_entry_size_bytes`
\ No newline at end of file
+soroban.restore-fprint-op.read-ledger-byte   | meter     | number of entry bytes in read entries
```
Comment (MonsieurNicolas) [inline docs/metrics.md:191]: fix wording, it's hard to understand (same for others in this group)

File: docs/metrics.md [hunk 6]
```diff
@@ -166,39 +166,49 @@ soroban.host-fn-op.read-entry                | meter     | number of entries rea
...
+soroban.restore-fprint-op.write-ledger-byte  | meter     | number of entry bytes in written entries
+soroban.ext-fprint-ttl-op.read-ledger-byte   | meter     | number of entry bytes in read entries
+soroban.ledger.cpu-insn                      | histogram | metered cpu instructions per ledger
+soroban.ledger.read-entry                    | histogram | number of entries read per ledger
+soroban.ledger.read-byte                     | histogram | number of entry bytes in read entries per ledger
+soroban.ledger.write-entry                   | histogram | number of entries written per ledger
+soroban.ledger.write-byte                    | histogram | number of entry bytes in written entries per ledger
+soroban.config.contract-max-rw-key-byte      | counter   | soroban config setting `contract_data_key_size_bytes`
```
Comment (MonsieurNicolas) [inline docs/metrics.md:199]: what is the point of exporting those soroban.config.* as metrics?

Comment (MonsieurNicolas) [reply docs/metrics.md:199]: (I agree they should be `counter` if we keep them)

**Review by MonsieurNicolas — CHANGES_REQUESTED** [review]

File: docs/metrics.md (hunk 6 above)
Comment (jayz22) [reply docs/metrics.md:199]: In order to display the limits on top of all of the resource consumption metrics (defined above). Without these we'll have to hardcode them on grafana and display them as variables, which is a static and inconvenient.

File: docs/metrics.md (hunk 5 above)
Comment (jayz22) [reply docs/metrics.md]: Fixed.

File: docs/metrics.md (hunk 4 above)
Comment (jayz22) [reply docs/metrics.md]: Ah yes, we didn't have it before (everything before was "meter"). Now we indeed don't need the count. I've reverted that commit (and keeps the "failure" metric).

File: docs/metrics.md (hunk 3 above)
Comment (jayz22) [reply docs/metrics.md]: I've cleaned up this section. Yes it is per-op.

File: docs/metrics.md (hunk 2 above)
Comment (jayz22) [reply docs/metrics.md]: Fixed.

File: docs/metrics.md (hunk 1 above)
Comment (jayz22) [reply docs/metrics.md]: Fixed.

**Comment by jayz22** [comment]
@MonsieurNicolas thanks for the review, and sorry about the confusion. Most of the confusion comes from "read" == both read and modified, and "write" == only modified, and "entry" == `LedgerEntry` which could be {ContractDataEntry, ContractCodeEntry}, and we also have "data" and "code" which are just those two variants.  
I've cleaned up the wording of the entire section, hope that makes it clear now.

**Review by MonsieurNicolas — APPROVED** [review]

**Comment by graydon** [comment]
r+ fed81d0

**Pull request merged**